### PR TITLE
Add preliminary support for go 1.16

### PR DESCRIPTION
**Pull request by sauyon**
Excludes treating embedded files as a variable initialization.

### CODE REVIEW

File: ql/test/library-tests/semmle/go/frameworks/StdlibTaintFlow/IoFs.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	{
+		source := newSource(4).(fs.FS)
+		fs.WalkDir(source, ".", func(_ string, d fs.DirEntry, _ error) error {
+			sink(4, d)
+			return nil
+		})
+	}
+	{
+		source := newSource(5).(fs.FS)
+		fs.WalkDir(source, ".", func(path string, _ fs.DirEntry, _ error) error {
```
Comment (owen-mc) [inline ql/test/library-tests/semmle/go/frameworks/StdlibTaintFlow/IoFs.go:37]: For one of the tests of `WalkDir`, how about using a function that's declared at file scope?

**Review by owen-mc — CHANGES_REQUESTED** [review]
- It would have been easier to review if this was split into separate commits for the different features being supported.
- What is the binary file `go1.16`?
- I expected `File.Read()` to be modelled, but I guess it's covered by `io.Reader.Read()`. It might be worth putting a test for it with the other `io/fs` tests.
- Can you add a test for the change you made to `Os.qll`? I can't find any tests for the rest of that predicate, so I accept this might be difficult.

File: ql/src/semmle/go/frameworks/stdlib/IoFs.qll [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+      input = inp and output = outp
+    }
+  }
+
+  private class WalkDirStep extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      //signature: func WalkDir(fsys FS, root string, fn WalkDirFunc) error
+      exists(DataFlow::CallNode call, FuncDef def |
```
Comment (smowton) [inline ql/src/semmle/go/frameworks/stdlib/IoFs.qll:43]: Worth a comment explaining that this is routing taint from WalkDir's arg0 to args 0 and 1 of the passed WalkDirFunc

**Comment by sauyon** [comment]
>     * It would have been easier to review if this was split into separate commits for the different features being supported.

It ended up mostly being the FS model so I didn't bother splitting it out, but I can do.

>     * What is the binary file `go1.16`?

A result of me running `go build` to test builds, oops.

>     * I expected `File.Read()` to be modelled, but I guess it's covered by `io.Reader.Read()`. It might be worth putting a test for it with the other `io/fs` tests.

:+1: 

>     * Can you add a test for the change you made to `Os.qll`? I can't find any tests for the rest of that predicate, so I accept this might be difficult.

Indeed, tests didn't exist, but it's probably not really an excuse for me not to add tests.

**Comment by smowton** [comment]
The go1.16 file is still there

**Comment by sauyon** [comment]
I haven't yet pushed, I'm adding more tests and adding a model for `File.Read`, because it turns out because it's an interface method we don't consider it to implement `io.Reader.Read`.

**Comment by sauyon** [comment]
Ok, everything should work now, including an additional (very confusing) change for a rename from `os.FileInfo` to `io/fs.FileInfo`.

File: ql/test/extractor-tests/go1.16/embed.expected [hunk 3]
```diff
@@ -0,0 +1 @@
+ERROR: Query module does not have any queries in scope. (embed.ql:1,1-1)
```
Comment (owen-mc) [inline ql/test/extractor-tests/go1.16/embed.expected:1]: You pushed a change before which fixed this error by adding something like `where exists(v.getDeclaration())`. Did that get lost somehow?

**Comment by gagliardetto** [comment]
@sauyon Did you use codebox, or that's handwritten following that code style?

**Comment by sauyon** [comment]
Handwritten, have yet to get `codebox` running, hah.

**Comment by gagliardetto** [comment]
Is it because of time constraints, or because it doesn't work?

**Comment by sauyon** [comment]
I tried to run it once and the server errored; haven't gotten to figuring out why yet.

**Comment by gagliardetto** [comment]
It could be used to diff the api of various packages between the current Go version and go1.16

File: ql/test/extractor-tests/go1.16/embed.expected (hunk 3 above)
Comment (sauyon) [reply ql/test/extractor-tests/go1.16/embed.expected:1]: Ugh. I think I pulled and git automatically rebased away some of my changes.

**Comment by sauyon** [comment]
> It could be used to diff the api of various packages between the current Go version and go1.16

They list the relevant changes in the change notes. The only reason I had issues with `FileInfo` is because apparently we don't generate `hasQualifiedName` sensibly for alias types.

**Comment by gagliardetto** [comment]
I'm not sure they go into extreme detail in the release notes. But then again, it might not be 100% important to have each and every function modeled.

**Review by owen-mc — APPROVED** [review]

**Comment by sauyon** [comment]
I've gone through and modeled a few new changes that appeared in the patch notes; probably a lesson not to rely too heavily on Go tip.

**Review by owen-mc — COMMENTED** [review]
Should `NopCloser` from `io/ioutil` be modeled in `Io.qll` now that it's been moved there?

I see you didn't delete `IoIoutil.qll`. Is it needed to support using older versions of Go?

**Comment by sauyon** [comment]
> Should `NopCloser` from `io/ioutil` be modeled in `Io.qll` now that it's been moved there?

Somehow missed this...

> I see you didn't delete `IoIoutil.qll`. Is it needed to support using older versions of Go?

It's deprecated and not removed (and won't be while Go is still in version 1).

**Review by owen-mc — APPROVED** [review]

**Review by smowton — APPROVED** [review]

File: ql/src/semmle/go/frameworks/stdlib/Io.qll [hunk 4]
```diff
@@ -61,6 +61,14 @@ module Io {
       // signature: func WriteString(w Writer, s string) (n int, err error)
       hasQualifiedName("io", "WriteString") and
       (inp.isParameter(1) and outp.isParameter(0))
+      or
+      // signature: func NopCloser(r io.Reader) io.ReadCloser
+      hasQualifiedName("io/ioutil", "NopCloser") and
```
Comment (owen-mc) [inline ql/src/semmle/go/frameworks/stdlib/Io.qll:66]:
```suggestion
      hasQualifiedName("io", "NopCloser") and
```

Comment (sauyon) [reply ql/src/semmle/go/frameworks/stdlib/Io.qll:66]: :facepalm:

**Comment by sauyon** [comment]
Well, at least the test I added worked.

**Comment by sauyon** [comment]
Hm. @owen-mc mind testing the OS X issue on your machine on Monday?

**Comment by sauyon** [comment]
I'd forgotten Chris also has a Mac, I've now solved the problem.

**Pull request merged**